### PR TITLE
[SPARK-44577][SQL] Fix INSERT BY NAME returns nonsensical error message

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
Fix INSERT BY NAME returns nonsensical error message on v1 datasource.
eg:
```scala
CREATE TABLE bug(c1 INT);

INSERT INTO bug BY NAME SELECT 1 AS c2;

==> Multi-part identifier cannot be empty.
```
After PR:
```scala
[INCOMPATIBLE_DATA_FOR_TABLE.CANNOT_FIND_DATA] Cannot write incompatible data for the table `spark_catalog`.`default`.`bug`: Cannot find data for the output column `c1`.
```

Also fixed the same issue when throwing other INCOMPATIBLE_DATA_FOR_TABLE type errors

### Why are the changes needed?
Fix the error msg nonsensical.

### Does this PR introduce _any_ user-facing change?
Yes, the error msg in v1 insert by name will be changed.

### How was this patch tested?
add new test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 1]
```diff
@@ -241,7 +241,7 @@ object TableOutputResolver {
         val extraCols = inputCols.filterNot(col => matchedCols.contains(col.name))
           .map(col => s"${toSQLId(col.name)}").mkString(", ")
         throw QueryCompilationErrors.incompatibleDataToTableExtraStructFieldsError(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:243]: I think we should throw `QueryCompilationErrors.cannotWriteTooManyColumnsToTableError` or `cannotWriteNotEnoughColumnsToTableError` here if `colPath` is empty.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 2]
```diff
@@ -263,16 +268,26 @@ object TableOutputResolver {
...
+        )
+      }
     } else if (inputCols.size < expectedCols.size) {
       val missingColsStr = expectedCols.takeRight(expectedCols.size - inputCols.size)
         .map(col => toSQLId(col.name))
         .mkString(", ")
-      throw QueryCompilationErrors.incompatibleDataToTableStructMissingFieldsError(
-        tableName, colPath.quoted, missingColsStr
-      )
+      if (colPath.isEmpty) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:283]: this appears three times, we can add a method for it.

**Comment by cloud-fan** [comment]
can we update the PR description for the new changes?

**Comment by cloud-fan** [comment]
```
[INSERT_COLUMN_ARITY_MISMATCH.NOT_ENOUGH_DATA_COLUMNS] Cannot write to `spark_catalog`.`default`.`bug`, the reason is not enough data columns:
Table columns: `c1`.
Data columns: `c2`.
```

For by-name insert, shall we just say "incompatible data columns"? The number of columns does match here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 3]
```diff
@@ -238,11 +238,17 @@ object TableOutputResolver {
 
     if (reordered.length == expectedCols.length) {
       if (matchedCols.size < inputCols.length) {
-        val extraCols = inputCols.filterNot(col => matchedCols.contains(col.name))
-          .map(col => s"${toSQLId(col.name)}").mkString(", ")
-        throw QueryCompilationErrors.incompatibleDataToTableExtraStructFieldsError(
-          tableName, colPath.quoted, extraCols
-        )
+        if (colPath.isEmpty) {
+          val cannotFindCol = expectedCols.filter(col => !matchedCols.contains(col.name)).head.name
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:242]: Can this be empty `expectedCols.filter(col => !matchedCols.contains(col.name))`? If so, `.head` can fail.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 4]
```diff
@@ -238,11 +238,17 @@ object TableOutputResolver {
...
+          throw QueryCompilationErrors.incompatibleDataToTableCannotFindDataError(tableName,
+            cannotFindCol)
+        } else {
+          val extraCols = inputCols.filterNot(col => matchedCols.contains(col.name))
+                  .map(col => s"${toSQLId(col.name)}").mkString(", ")
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:247]: Fix indentation

**Comment by cloud-fan** [comment]
@MaxGekk does this PR align with your plan of unifying the assignment errors?

**Comment by MaxGekk** [comment]
@cloud-fan In my PR https://github.com/apache/spark/pull/42393, I am going to touch non-`BY NAME` functionality so far.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 5]
```diff
@@ -238,11 +238,17 @@ object TableOutputResolver {
...
-          tableName, colPath.quoted, extraCols
-        )
+        if (colPath.isEmpty) {
+          val cannotFindCol = expectedCols.filter(col => !matchedCols.contains(col.name)).head.name
+          throw QueryCompilationErrors.incompatibleDataToTableCannotFindDataError(tableName,
+            cannotFindCol)
+        } else {
+          val extraCols = inputCols.filterNot(col => matchedCols.contains(col.name))
+            .map(col => s"${toSQLId(col.name)}").mkString(", ")
+          throw QueryCompilationErrors.incompatibleDataToTableExtraStructFieldsError(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:248]: so top-level column and nested field have different error classes?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:248]: extra field? I think it's for missing column/field here?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:248]: so the behavior is inconsistent between top-level columns and struct fields?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:248]: I still don't quite understand the code here. If it's for extra columns/fields, why do we throw an error for top-level columns?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:248]: My question is simple: this is the branch of `if (matchedCols.size < inputCols.length) {` which means extra columns, why do we throw `incompatibleDataToTableCannotFindDataError` for top-level columns which seems to be missing columns?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:248]: `table: [a, b, c], input [a, c, d]`
Extra column and missing column are orthogonal, they can both occur, or only one of them. Can you elaborate on the error checking logic here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala [hunk 6]
```diff
@@ -238,11 +238,17 @@ object TableOutputResolver {
...
     if (reordered.length == expectedCols.length) {
       if (matchedCols.size < inputCols.length) {
-        val extraCols = inputCols.filterNot(col => matchedCols.contains(col.name))
-          .map(col => s"${toSQLId(col.name)}").mkString(", ")
-        throw QueryCompilationErrors.incompatibleDataToTableExtraStructFieldsError(
-          tableName, colPath.quoted, extraCols
-        )
+        if (colPath.isEmpty) {
+          val cannotFindCol = expectedCols.filter(col => !matchedCols.contains(col.name)).head.name
+          throw QueryCompilationErrors.incompatibleDataToTableCannotFindDataError(tableName,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:243]: when we reach here, why it means missing col for top-level columns, but extra column for nested fields?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:243]: where do we check missing cols error for nested fields?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:243]: The code here is so convoluted...

`if (reordered.length == expectedCols.length)` I think this means there is no missing col? For every table column, we can find a match in the input columns.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:243]: >  eg: table [x, y, z], inputcol: [x, y, k],

If `z` has default value, I think it means extra column error?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala:243]: missing column error should only be thrown if that missing column does not have a default value.

File: common/utils/src/main/resources/error/error-classes.json [hunk 7]
```diff
@@ -1025,6 +1025,11 @@
           "Cannot safely cast <colName> <srcType> to <targetType>."
         ]
       },
+      "EXTRA_FIELDS" : {
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-classes.json:1028]: `EXTRAC_COLUMNS`?

File: common/utils/src/main/resources/error/error-classes.json [hunk 8]
```diff
@@ -1025,6 +1025,11 @@
...
+        "message" : [
+          "Cannot write extra fields <colName>."
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-classes.json:1030]: it's one column or multiple columns?

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5!

**Pull request closed without merging**